src/System.IO.Pipelines.Networking.Sockets/SocketConnection.cs

```diff
@@ -167,30 +179,40 @@ internal static void RecycleSocketAsyncEventArgs(SocketAsyncEventArgs args)
                 }
             }
         }
-        /// <summary>
-        /// Releases all resources owned by the connection
-        /// </summary>
-        protected virtual void Dispose(bool disposing)
+
+        public async Task DisposeAsync()
         {
-            if (disposing)
+            if (!_disposed)
             {
+                _disposed = true;
                 _stopping = true;
                 _output.Reader.CancelPendingRead();
 
-                Task.WaitAll(_sendTask, _receiveTask);
+                await Task.WhenAll(_sendTask, _receiveTask);
 
                 _output.Writer.Complete();
                 _input.Reader.Complete();
 
-                GC.SuppressFinalize(this);
-
                 _socket?.Dispose();
                 _socket = null;
                 if (_ownsFactory) { _factory?.Dispose(); }
                 _factory = null;
             }
         }
 
+        /// <summary>
+        /// Releases all resources owned by the connection
+        /// </summary>
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                GC.SuppressFinalize(this);
```

Comment:
Why even implement the disposing patter if this is the only virtual method in the class? The pattern also recommends putting the call to SuppressFinalize in the non-virtual Dispose() method, presumably so it can't be overridden.